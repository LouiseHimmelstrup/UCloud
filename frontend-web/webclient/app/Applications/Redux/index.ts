import * as ViewObject from "./ViewObject";
import * as ViewReducer from "./ViewReducer";
import * as ApplicationObject from "./ApplicationsObject";
import * as ApplicationReducer from "./ApplicationsReducer";

export type Reducers = ViewReducer.Reducer & ApplicationReducer.Reducer;
export type Objects = ViewObject.Wrapper & ApplicationObject.Wrapper;

export function init(): Objects {
    return { 
        ...ViewObject.init(),
        ...ApplicationObject.init()
    }
}

export const reducers: Reducers = {
    applicationView: ViewReducer.default,
    applicationsBrowse: ApplicationReducer.default
};