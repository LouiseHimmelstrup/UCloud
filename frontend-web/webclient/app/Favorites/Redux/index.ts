import * as FavoritesObject from "./FavoriteObject";
import * as FavoritesReducer from "./FavoritesReducer";

export type Reducers = FavoritesReducer.Reducer;
export type Objects = FavoritesObject.Wrapper;

export function init(): Objects {
    return {
        ...FavoritesObject.init()
    }
}

export const reducers: Reducers = {
    favorites: FavoritesReducer.default
}