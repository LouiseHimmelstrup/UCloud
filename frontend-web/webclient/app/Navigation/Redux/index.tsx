import {PRODUCT_NAME} from "../../../site.config.json";
import {SetLoadingAction} from "@/Types";
import {useDispatch} from "react-redux";
import {useEffect} from "react";
import {initStatus, StatusReduxObject} from "@/DefaultObjects";

export type Index = UpdatePageTitleAction | SetLoading;

export type UpdatePageTitleAction = PayloadAction<typeof UPDATE_PAGE_TITLE, {title: string}>;
/**
 * Sets the title of the window. Stores in the redux store as well
 * @param {string} title the title to be set
 */
export const updatePageTitle = (title: string): UpdatePageTitleAction => ({
    type: UPDATE_PAGE_TITLE,
    payload: {title}
});

type SetLoading = SetLoadingAction<typeof SET_STATUS_LOADING>;
export const setLoading = (loading: boolean): SetLoading => ({
    type: SET_STATUS_LOADING,
    payload: {loading}
});

export interface SetStatusLoading {
    setLoading: (loading: boolean) => void;
}

export function useTitle(title: string): void {
    const dispatch = useDispatch();
    useEffect(() => {
        dispatch(updatePageTitle(title));
        return () => {
            dispatch(updatePageTitle(""));
        };
    }, [title]);
}

export function useLoading(loading: boolean): void {
    const dispatch = useDispatch();
    useEffect(() => {
        dispatch(setLoading(loading));
    }, [loading]);
}

export const UPDATE_PAGE_TITLE = "UPDATE_PAGE_TITLE";
export const SET_STATUS_LOADING = "SET_STATUS_LOADING";

export const statusReducer = (state: StatusReduxObject = initStatus(), action: Index): StatusReduxObject => {
    switch (action.type) {
        case UPDATE_PAGE_TITLE:
            document.title = `${PRODUCT_NAME} | ${action.payload.title}`;
            return {...state, ...action.payload};
        case SET_STATUS_LOADING:
            return {...state, ...action.payload};
        default: {
            return state;
        }
    }
};

