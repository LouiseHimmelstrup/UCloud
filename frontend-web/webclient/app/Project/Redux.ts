import {File} from "Files";
import {Action} from "redux";
import {Dictionary, Page} from "Types";
import {getParentPath, resolvePath} from "Utilities/FileUtilities";

export interface State {
    project?: string;
    files?: Dictionary<ProjectFilesCacheEntry>;
}

export interface ProjectFilesCacheEntry {
    project: string;
    path: string;
    files: Page<File>;
}

export const initialState = {project: getStoredProject() ?? undefined};

interface SetProjectAction extends Action<"SET_PROJECT"> {
    project?: string;
}

export function setProjectAction(project?: string): SetProjectAction {
    return {project, type: "SET_PROJECT"};
}

interface UpdateProjectFileAction extends Action<"UPDATE_PROJECT_FILE"> {
    newFile: File;
}

export function updateProjectFileAction(newFile: File): UpdateProjectFileAction {
    return {newFile, type: "UPDATE_PROJECT_FILE"};
}

type CacheProjectFilesAction = Action<"CACHE_PROJECT_FILES"> & ProjectFilesCacheEntry;

export function cacheProjectFilesAction(entry: ProjectFilesCacheEntry): CacheProjectFilesAction {
    return {...entry, type: "CACHE_PROJECT_FILES"};
}

type ClearDirtFlagAction = Action<"CACHE_PROJECT_FILES_CLEAR_DIRTY">;

export function clearProjectFilesDirtyFlag(): ClearDirtFlagAction {
    return {type: "CACHE_PROJECT_FILES_CLEAR_DIRTY"};
}

type ProjectAction = SetProjectAction | CacheProjectFilesAction | UpdateProjectFileAction | ClearDirtFlagAction;

export const reducer = (state: State = initialState, action: ProjectAction) => {
    switch (action.type) {
        case "SET_PROJECT": {
            setStoredProject(action.project ?? null);
            return {...state, project: action.project};
        }

        case "CACHE_PROJECT_FILES": {
            const existingDirectoryOrInitial = state.files ?? {};
            const files = {...existingDirectoryOrInitial};
            files[action.path] = action;
            return {...state, files};
        }

        case "UPDATE_PROJECT_FILE": {
            const parentPath = resolvePath(getParentPath(action.newFile.path));
            const newFile = {...action.newFile, frontend: {dirty: true}};
            const cachedParent: ProjectFilesCacheEntry = state.files?.[parentPath] ??
                {
                    project: state.project!,
                    path: parentPath,
                    files: {
                        items: [newFile],
                        pageNumber: 0,
                        itemsPerPage: 50,
                        itemsInTotal: 1,
                        pagesInTotal: 1
                    }
                };

            const directoryCopy = cachedParent.files.items.slice();
            for (let i = 0; i < directoryCopy.length; i++) {
                const file = directoryCopy[i];
                if (file.path === newFile.path) {
                    directoryCopy[i] = newFile;
                    break;
                }
            }

            const files = {...state.files};
            files[parentPath] = {...cachedParent, files: {...cachedParent.files, items: directoryCopy}};
            return {...state, files};
        }

        case "CACHE_PROJECT_FILES_CLEAR_DIRTY": {
            const files: Dictionary<ProjectFilesCacheEntry> = {...state.files};
            Object.keys(files).forEach(key => {
                const current = files[key];
                files[key] = {
                    ...current,
                    files: {
                        ...current.files,
                        items: current.files.items.map(it => ({...it, frontend: undefined}))
                    }
                };
            });

            return {...state, files};
        }

        default:
            return state;
    }
};

function getStoredProject(): string | null {
    return window.localStorage.getItem("project") ?? null;
}

function setStoredProject(value: string | null) {
    if (value === null) {
        window.localStorage.removeItem("project");
    } else {
        window.localStorage.setItem("project", value);
    }
}

