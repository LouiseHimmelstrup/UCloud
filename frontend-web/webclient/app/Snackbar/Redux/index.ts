import snackbar from "./SnackbarsReducer";
import { Snack } from "Snackbar/Snackbars";

export const init = (): Wrapper => ({
    snackbar: {
        snackbar: [] as Snack[],
        nextId: 0
    }
});

export const reducers = {
    snackbar
};

export interface Wrapper {
    snackbar: Object
}

export interface Object {
    snackbar: Snack[]
    nextId: number
}