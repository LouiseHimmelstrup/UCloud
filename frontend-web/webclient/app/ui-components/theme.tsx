const createMediaQuery = (n: string | number) => `@media screen and (min-width:${n})`

const addAliases = (arr: any, aliases: any[]) =>
  aliases.forEach((key, i) =>
    Object.defineProperty(arr, key, {
      enumerable: false,
      get() {
        return this[i]
      }
    })
  )

// export const breakpoints = [32, 40, 48, 64, 80].map(n => n + 'em')
export const breakpoints = [512, 640, 768, 1024, 1280].map(n => n + 'px')
const aliases = ['xs', 'sm', 'md', 'lg', 'xl']
export const responsiveBP = { xs: 512-1, sm: 640-1, md: 768-1, lg: 1024-1, xl: 1280-1 } 

export const mediaQueries = breakpoints.map(createMediaQuery)
addAliases(breakpoints, aliases)
addAliases(mediaQueries, aliases)

export const space = [0, 4, 8, 16, 32, 64, 128]

export const fontFamily = `'IBM Plex Sans', sans-serif`

export const fontSizes = [10, 14, 16, 20, 24, 32, 40, 56, 72]

export const medium = 300
export const bold = 700
export const regular = 400

// styled-system's `fontWeight` function can hook into the `fontWeights` object
export const fontWeights = {
  medium,
  bold,
  regular
}

export const lineHeights = {
  standard: 1.5,
  display: 1.25
}

const letterSpacings = {
  normal: 'normal',
  caps: '0.025em'
}

export const textStyles = {
  display8: {
    fontSize: fontSizes[8] + 'px',
    fontWeight: fontWeights.bold,
    lineHeight: lineHeights.display
  },
  display7: {
    fontSize: fontSizes[7] + 'px',
    fontWeight: fontWeights.bold,
    lineHeight: lineHeights.display
  },
  display6: {
    fontSize: fontSizes[6] + 'px',
    fontWeight: fontWeights.bold,
    lineHeight: lineHeights.display
  },
  display5: {
    fontSize: fontSizes[5] + 'px',
    fontWeight: fontWeights.bold,
    lineHeight: lineHeights.display
  },
  display4: {
    fontSize: fontSizes[4] + 'px',
    fontWeight: fontWeights.bold,
    lineHeight: lineHeights.display
  },
  display3: {
    fontSize: fontSizes[3] + 'px',
    fontWeight: fontWeights.bold,
    lineHeight: lineHeights.display
  },
  display2: {
    fontSize: fontSizes[2] + 'px',
    fontWeight: fontWeights.bold,
    lineHeight: lineHeights.display
  },
  display1: {
    fontSize: fontSizes[1] + 'px',
    fontWeight: fontWeights.bold,
    lineHeight: lineHeights.display
  },
  display0: {
    fontSize: fontSizes[0] + 'px',
    fontWeight: fontWeights.bold,
    lineHeight: lineHeights.display,
    letterSpacing: letterSpacings.caps,
    textTransform: 'uppercase'
  },
  body2: {
    fontSize: fontSizes[2] + 'px',
    fontWeight: fontWeights.medium,
    lineHeight: lineHeights.standard
  },
  body1: {
    fontSize: fontSizes[1] + 'px',
    fontWeight: fontWeights.medium,
    lineHeight: lineHeights.standard
  },
  body0: {
    fontSize: fontSizes[0] + 'px',
    fontWeight: fontWeights.medium,
    lineHeight: lineHeights.standard
  }
}


// color palette
const black = "#000";
const white = "#fff";
const textBlack = "#1e252e"
//// Gray
const lightGray = "#ebeff3";
const midGray = "#c9d3df";
const gray = "#8393A7";
const darkGray = "#53657d";
//// Blue
const lightBlue = "#f0f6ff";
const lightBlue2 = "#cdf";
const blue = "#0055d5"; 
// const blue = "#007bff"; 
const darkBlue = "#049";
//// Green
const lightGreen = "#00ff77";
const green = "#00C05A"; // secondary
const darkGreen = "#00823c";
//// Red
const lightRed = "#fcc";
const red = "#c00"; // secondary
const darkRed = "#800";
//// Orange
const orange = "#fa0"; // secondary
const darkOrange = "#a50";
//// Purple
const lightPurple = "#ecf";
const purple = "#70b"; // secondary

// Colors in the array come in 3 shades: light, medium , dark
// last color is for logo centers only
const appColors = [
  //["#0096ff", "#043eff"], // blue
  ["#F7D06A", "#E98C33", "#C46927"], // gold
  ["#EC6F8E", "#C75480", "#AA2457"], // salmon
  ["#B8D1E3", "#7C8DB3", "#5B698C"], // silver
  ["#83D8F9", "#3F80F6", "#2951BE"], // blue
  ["#AE83CF", "#9065D1", "#68449E"], // violet
  ["#E392CC", "#E2689D", "#B33B6D"], // pink
  ["#ECB08C", "#EA7B4B", "#BC4F33"], // bronze
  ["#90DCA1", "#69C97D", "#4D9161"], // green
  ["#F3B576", "#B77D50", "#7C4C3C"], // brown
  ["#D57AC5", "#E439C9", "#A1328F"], // purple
  ["#98E0F9", "#53A5F5", "#3E79C0"], // lightblue
  ["#DC6AA6", "#C62A5A", "#AA2457"], // red
  ["#C9D3DF", "#8393A7", "#53657D"], // gray colors from the theme
];

// Color Themes
// should use colors from the palette above
//// Light color theme
// const text = "#001833";
const text = textBlack;
const textHighlight = blue;
const headerText = lightGray;
const headerBg = blue;
const headerIconColor = lightGray
const headerIconColor2 = midGray;
// const borderGray = "#d1d6db";
const borderGray = midGray; //used for borders of cards, pagination, sidebar
//const paginationHoverColor = "#f7f7f7"
const paginationHoverColor = lightBlue;
const paginationDisabled = lightGray;
// Icons colors
const iconColor = darkGray;
const iconColor2 = gray;
const FtIconColor = lightGray;
const FtIconColor2 = gray;
const spinnerColor = blue;
// File table colors
const tableRowHighlight = lightBlue;


const colors = {
  black,
  white,
  lightGray,
  midGray,
  gray,
  darkGray,
  blue,
  lightBlue,
  lightBlue2,
  darkBlue,
  green,
  lightGreen,
  darkGreen,
  red,
  lightRed,
  darkRed,
  orange,
  darkOrange,
  purple,
  lightPurple,
  text,
  textHighlight,
  headerText,
  headerBg,
  headerIconColor,
  headerIconColor2,
  borderGray,
  paginationHoverColor,
  paginationDisabled,
  iconColor,
  iconColor2,
  FtIconColor,
  FtIconColor2,
  spinnerColor,
  tableRowHighlight,
}

export type ThemeColor = keyof typeof colors;


export { colors }

export const colorStyles = {
  whiteOnText: {
    color: colors.white,
    backgroundColor: colors.text
  },
  whiteOnGray: {
    color: colors.white,
    backgroundColor: colors.gray
  },
  textOnLightGray: {
    color: colors.text,
    backgroundColor: colors.lightGray
  },
  whiteOnBlue: {
    color: colors.white,
    backgroundColor: colors.blue
  },
  blueOnLightBlue: {
    color: colors.blue,
    backgroundColor: colors.lightBlue
  },
  whiteOnGreen: {
    color: colors.white,
    backgroundColor: colors.green
  },
  greenOnLightGreen: {
    color: colors.green,
    backgroundColor: colors.lightGreen
  },
  whiteOnRed: {
    color: colors.white,
    backgroundColor: colors.red
  },
  redOnLightRed: {
    color: colors.red,
    backgroundColor: colors.lightRed
  },
  textOnOrange: {
    color: colors.text,
    backgroundColor: colors.orange
  },
  whiteOnPurple: {
    color: colors.white,
    backgroundColor: colors.purple
  },
  purpleOnLightPurple: {
    color: colors.purple,
    backgroundColor: colors.lightPurple
  },
  textOnWhite: {
    color: colors.text,
    backgroundColor: colors.white
  },
  grayOnWhite: {
    color: colors.gray,
    backgroundColor: colors.white
  },
  blueOnWhite: {
    color: colors.blue,
    backgroundColor: colors.white
  },
  greenOnWhite: {
    color: colors.green,
    backgroundColor: colors.white
  },
  redOnWhite: {
    color: colors.red,
    backgroundColor: colors.white
  },
  purpleOnWhite: {
    color: colors.purple,
    backgroundColor: colors.white
  },
  whiteOnDarkOrange: {
    color: colors.white,
    backgroundColor: colors.darkOrange
  },
  // info: textOnLightGray
  info: {
    color: colors.text,
    backgroundColor: colors.lightGray
  },
  // success: whiteOnGreen
  success: {
    color: colors.white,
    backgroundColor: colors.green
  },
  //warning: textOnOrange
  warning: {
    color: colors.text,
    backgroundColor: colors.orange
  },
  // danger: whiteOnRed
  danger: {
    color: colors.white,
    backgroundColor: colors.red
  }
}

// styled-system's `borderRadius` function can hook into the `radii` object/array
export const radii = [0, 2, 6]
export const radius = '5px'

export const maxContainerWidth = '1280px'

// boxShadows: styled-systems hooks into shadows
export const shadows = [
  `0 0 2px 0 rgba(0,0,0,.08),0 1px 4px 0 rgba(0,0,0,.16)`,
  `0 0 2px 0 rgba(0,0,0,.08),0 2px 8px 0 rgba(0,0,0,.16)`,
  `0 0 2px 0 rgba(0,0,0,.08),0 4px 16px 0 rgba(0,0,0,.16)`,
  `0 0 2px 0 rgba(0,0,0,.08),0 8px 32px 0 rgba(0,0,0,.16)`
]
const BoxShadowsAliases = ['sm', 'md', 'lg', 'xl'];
addAliases(shadows, BoxShadowsAliases);

// animation duration
export const duration = {
  fast: `150ms`,
  normal: `300ms`,
  slow: `450ms`,
  slowest: `600ms`
}

// animation easing curves
const easeInOut = 'cubic-bezier(0.5, 0, 0.25, 1)'
const easeOut = 'cubic-bezier(0, 0, 0.25, 1)'
const easeIn = 'cubic-bezier(0.5, 0, 1, 1)'

const timingFunctions = {
  easeInOut,
  easeOut,
  easeIn
}

// animation delay
const transitionDelays = {
  small: `60ms`,
  medium: `160ms`,
  large: `260ms`,
  xLarge: `360ms`
}

const theme = {
  breakpoints,
  mediaQueries,
  space,
  fontFamily,
  fontSizes,
  fontWeights,
  lineHeights,
  letterSpacings,
  regular,
  bold,
  textStyles,
  colors,
  colorStyles,
  appColors,
  radii,
  radius,
  shadows,
  maxContainerWidth,
  duration,
  timingFunctions,
  transitionDelays
}

export type Theme = typeof theme

export default theme
