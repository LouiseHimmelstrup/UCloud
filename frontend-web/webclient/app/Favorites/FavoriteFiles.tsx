import * as React from "react";
import { List } from "Pagination/List";
import { MainContainer } from "MainContainer/MainContainer";
import { connect } from "react-redux";
import { favoriteFileAsync, AllFileOperations } from "Utilities/FileUtilities";
import { Cloud } from "Authentication/SDUCloudObject";
import { Type as ReduxType } from "./Redux/FavoriteObject";
import { SortBy, SortOrder, File } from "Files";
import { emptyPage, ReduxObject } from "DefaultObjects";
import * as Heading from "ui-components/Heading";
import { FilesTable, FileOptions } from "Files/FilesTable";
import { Dispatch } from "redux";
import { setRefreshFunction } from "Navigation/Redux/HeaderActions";
import { fetchFavorites, setLoading, receiveFavorites } from "./Redux/FavoritesActions";
import { MasterCheckbox } from "UtilityComponents";
import { Page } from "Types";

class FavoriteFiles extends React.Component<FavoritesOperations & ReduxType & { header: any }> {
    constructor(props: Readonly<FavoritesOperations & ReduxType & { header: any }>) {
        super(props);
        props.fetchFileFavorites(emptyPage.pageNumber, emptyPage.itemsPerPage);
        props.setRefresh(() => props.fetchFileFavorites(emptyPage.pageNumber, emptyPage.itemsPerPage));
    }

    public componentWillUnmount() {
        this.props.setRefresh();
    }

    public componentWillReceiveProps(nextProps) {
        const { setRefresh, page, fetchFileFavorites } = nextProps;
        setRefresh(() => fetchFileFavorites(page.pageNumber, page.itemsPerPage));
    }

    // FIXME: Make action instead
    checkFile(checked: boolean, path: string) {
        this.props.page.items.find(it => it.path === path)!.isChecked = checked;
        this.props.receiveFavorites(this.props.page);
    }

    // FIXME: Make action instead
    checkAll(checked: boolean) {
        this.props.page.items.forEach(it => it.isChecked = checked);
        this.props.receiveFavorites(this.props.page);
    }

    public render() {
        const fileoperations = AllFileOperations({ stateless: true, setLoading: () => this.setState(() => ({ loading: true })) });
        const { page } = this.props;
        const { pageNumber, itemsPerPage } = page;
        const selectedFiles = page.items.filter(it => it.isChecked);
        return (<MainContainer
            header={this.props.header}
            main={<List
                page={this.props.page}
                loading={this.props.loading}
                customEmptyPage={<Heading.h2>You have no favorites</Heading.h2>}
                onPageChanged={pageNumber => this.props.fetchFileFavorites(pageNumber, itemsPerPage)}
                pageRenderer={page =>
                    <FilesTable
                        onFavoriteFile={async files => (await favoriteFileAsync(files[0], Cloud), this.props.fetchFileFavorites(pageNumber, itemsPerPage))}
                        fileOperations={fileoperations}
                        files={page.items}
                        sortBy={SortBy.PATH}
                        sortOrder={SortOrder.DESCENDING}
                        /* FIXME: ADD */
                        sortFiles={() => undefined}
                        /* FIXME: ADD */
                        refetchFiles={() => undefined}
                        sortingColumns={[SortBy.MODIFIED_AT, SortBy.SIZE]}
                        /* FIXME Fix */
                        onCheckFile={(checked, file) => this.checkFile(checked, file.path)}
                        masterCheckbox={
                            <MasterCheckbox
                                onClick={check => this.checkAll(check)}
                                checked={page.items.length === selectedFiles.length && page.items.length > 0}
                            />}
                    />}
            />
            }
            sidebar={
                <FileOptions files={this.props.page.items.filter(it => it.isChecked)} fileOperations={fileoperations} />
            }
        />)
    }
}

interface FavoritesOperations {
    setRefresh: (refresh?: () => void) => void
    fetchFileFavorites: (pageNumber: number, itemsPerPage: number) => void
    receiveFavorites: (page: Page<File>) => void
}

const mapDispatchToProps = (dispatch: Dispatch): FavoritesOperations => ({
    setRefresh: refresh => dispatch(setRefreshFunction(refresh)),
    fetchFileFavorites: async (pageNumber, itemsPerPage) => {
        setLoading(true);
        dispatch(await fetchFavorites(pageNumber, itemsPerPage))
        setLoading(false);
    },
    receiveFavorites: page => dispatch(receiveFavorites(page))
});

const mapStateToProps = ({ favorites }: ReduxObject): ReduxType & { checkedCount: number } => ({ ...favorites, checkedCount: favorites.page.items.filter(it => it.isChecked).length });

export default connect(mapStateToProps, mapDispatchToProps)(FavoriteFiles);