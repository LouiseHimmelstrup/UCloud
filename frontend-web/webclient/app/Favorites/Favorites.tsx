import * as React from "react";
import { connect } from "react-redux";
import Installed from "Applications/Installed";
import { SearchOptions, SelectableText } from "Search/Search";
import { prettierString } from "UtilityFunctions";
import FavoriteFiles from "./FavoriteFiles";
import { Dispatch } from "redux";
import { updatePageTitle, setActivePage } from "Navigation/Redux/StatusActions";
import { ReduxObject } from "DefaultObjects";
import { setFavoritesShown } from "./Redux/FavoritesActions";
import { SidebarPages } from "ui-components/Sidebar";
import { History } from "history";

export enum FavoriteType {
    FILES = "FILES",
    APPLICATIONS = "APPLICATIONS"
}

const Favorites = (props: FavoritesStateProps & FavoritesOperations & { history: History }) => {
    React.useEffect(() => {
        props.setPageTitle();
        props.setActivePage(SidebarPages.Favorites);
        return () => props.setActivePage(SidebarPages.None)
    }, []);

    const { shown, setShown } = props;
    const header = (<SearchOptions>
        {Object.keys(FavoriteType).map((it: FavoriteType) =>
            <SelectableText cursor="pointer" key={it} onClick={() => setShown(it)} mr="1em" selected={it === shown}>
                {prettierString(it)}
            </SelectableText>
        )}
    </SearchOptions>);
    if (shown === FavoriteType.FILES) {
        return (<FavoriteFiles header={header} history={props.history} />);
    } else {
        return (<Installed header={header} />)
    }
}

interface FavoritesStateProps { shown: FavoriteType }

const mapStateToProps = (state: ReduxObject): FavoritesStateProps => ({ shown: state.favorites.shown })

interface FavoritesOperations {
    setPageTitle: () => void
    setShown: (type: FavoriteType) => void
    setActivePage: (page: SidebarPages) => void
}

const mapDispatchToProps = (dispatch: Dispatch): FavoritesOperations => ({
    setPageTitle: () => dispatch(updatePageTitle("Favorites")),
    setActivePage: page => dispatch(setActivePage(page)),
    setShown: type => dispatch(setFavoritesShown(type))
});

export default connect(mapStateToProps, mapDispatchToProps)(Favorites);