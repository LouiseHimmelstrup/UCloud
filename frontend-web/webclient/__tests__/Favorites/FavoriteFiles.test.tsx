import * as React from "react";
import FavoriteFiles from "../../app/Favorites/FavoriteFiles";
import { create } from "react-test-renderer";
import { createMemoryHistory } from "history";
import { configureStore } from "../../app/Utilities/ReduxUtilities";
import * as FavoritesRedux from "../../app/Favorites/Redux";
import { Provider } from "react-redux";
import { createResponsiveStateReducer, responsiveStoreEnhancer } from "redux-responsive";
import theme, { responsiveBP } from "../../app/ui-components/theme";
import { ThemeProvider } from "styled-components";

const store = configureStore({
    ...FavoritesRedux.init(),
}, {
    ...FavoritesRedux.reducers,
    responsive: createResponsiveStateReducer(
        responsiveBP,
        { infinity: "xxl" })
}, responsiveStoreEnhancer)

describe("Favorite Files", () => {
    test.skip("Mount", () => {
        // The call to the backend doesn't fail, overwriting emptyPage with value 1
        expect(create(
            <Provider store={store}>
                <ThemeProvider theme={theme}>
                    <FavoriteFiles
                        header={null}
                        history={createMemoryHistory()}
                    />
                </ThemeProvider>
            </Provider>).toJSON()).toMatchSnapshot();
    });
});