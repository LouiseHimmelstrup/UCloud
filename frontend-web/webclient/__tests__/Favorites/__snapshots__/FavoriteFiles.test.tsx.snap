// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Favorite Files Mount 1`] = `
<div
  className="sc-bdVaJa fENTec"
  cursor="auto"
>
  <div
    className="sc-bdVaJa sc-jWBwVP hXrphb"
  >
    <div
      className="sc-bdVaJa sc-bwzfXH sc-cvbbAY cqZRxn"
      cursor="auto"
      height="100%"
      width={240}
    />
  </div>
  <div
    className="sc-bdVaJa iEzrdZ"
    cursor="auto"
  >
    <div
      className="sc-bdVaJa sc-jWBwVP jJhLPe"
    />
    <div
      className="sc-cmTdod cpzqWF"
      height={24}
      width={24}
    >
      <div
        className="sc-jwKygS sc-btzYZH deBJuJ"
      />
      <div
        className="sc-jwKygS sc-lhVmIH DJpLr"
      />
      <div
        className="sc-jwKygS sc-bYSBpT bBpxRp"
      />
      <div
        className="sc-jwKygS sc-elJkPf eimgCQ"
      />
      <div
        className="sc-jwKygS sc-jtRfpW lbMlBI"
      />
      <div
        className="sc-jwKygS sc-kTUwUJ hRcYEv"
      />
      <div
        className="sc-jwKygS sc-dqBHgY rdbit"
      />
      <div
        className="sc-jwKygS sc-gxMtzJ hPYjLZ"
      />
      <div
        className="sc-jwKygS sc-dfVpRl gSTqeV"
      />
    </div>
  </div>
</div>
`;
